File: python/sorts/merge_sort.py
from typing import List
def merge(arr:List[int], left:int, mid:int, right:int)->None: #2nd flavour
    arr1, arr2 = arr[left:mid+1], arr[mid+1:right+1]
    n, m = len(arr1), len(arr2)
    i,j,k = 0, 0, left
    while i != n and j != m:
        if arr1[i] <= arr2[j]:
            arr[k] = arr1[i]
            k+=1
            i+=1
        else:
            arr[k] = arr2[j]
            k+=1
            j+=1
    while j < m:
        arr[k] = arr2[j]
        k+=1
        j+=1
    while i < n:
        arr[k] = arr1[i]
        k+=1
        i+=1

def mergeSort(arr:List[int], l:int, r:int)->None:
    if l >= r:
        return
    m = (l + r) // 2
    mergeSort(arr, l, m)
    mergeSort(arr, m+1, r)
    merge(arr, l, m , r)
        
arr = [12, 11, 13, 5, 6, 7] 
n = len(arr)
print(arr)
mergeSort(arr, 0, n-1)
print(arr)

File: python/sorts/merge_sort.cpp
#include<iostream>
#include<vector>
using namespace std;
void input(vector<int>&nums){
	/*
	add the user-defined values in the vector of ints
	input: vector<int>
	output: void
	*/
   for(auto &i : nums){
       cin >> i;
   }
}
void display(const vector<int>nums){
	/*
	display the vector of ints
	input: const vector<int> 
	output: void
	*/
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
void merge(vector<int>&nums, int l, int m, int r){
	/*
	merge the two arrays from left index to mid and mid +1 index to right index 
	input: vector<int>, left index: int, mid index: int and right index: int 
	output: void
	*/
	int i, j, k;
	i = k = l;
	j = m + 1;
	vector<int>temp(nums.size());
	
	while(i <= m && j <= r){
		if(nums[i] <= nums[j]){
			temp[k++] = nums[i++];
		} else{
			temp[k++] = nums[j++];			
		}		
	}
	while( i <= m){
		temp[k++] = nums[i++];
	}
	while (j <= r){
		temp[k++] = nums[j++];
	}
	
	for(int p = l; p <= r; p++){
		nums[p] = temp[p];
	}
	
}
void mergeSort(vector<int>&nums, int l, int r){
	/*
	sort the array ascendingly 
	input: take the vector of ints, left index: int, right index: int
	output: void
	*/
	if (l < r){
		int mid = (l + r) / 2;
		mergeSort(nums, l, mid);
		mergeSort(nums, mid+1, r);

		merge(nums,l, mid, r);
	}
}

int main()
{
    int n;
    cin >> n; 
    vector<int>nums(n); 
    input(nums);
    display(nums);
    mergeSort(nums,0, nums.size() -1 );
	display(nums);
    return 0;
}